src/core/config/crypto.h

```diff
@@ -0,0 +1,100 @@
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef CONFIG_CRYPTO_H_
+#define CONFIG_CRYPTO_H_
+
+/**
+ * @def OPENTHREAD_CONFIG_CRYPTO_LIB
+ *
+ * Selects the crypto backend library for OpenThread.
+ *
+ * There are several options available
+ * - @sa OPENTHREAD_CONFIG_CRYPTO_LIB_MBEDTLS_2
+ * - @sa OPENTHREAD_CONFIG_CRYPTO_LIB_PSA
+ * - @sa OPENTHREAD_CONFIG_CRYPTO_LIB_PLATFORM
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_CRYPTO_LIB
+#define OPENTHREAD_CONFIG_CRYPTO_LIB OPENTHREAD_CONFIG_CRYPTO_LIB_MBEDTLS_2
+#endif
+
+/** Use mbedtls v2.x as crypto library */
+#define OPENTHREAD_CONFIG_CRYPTO_LIB_MBEDTLS_2 0
+/** Use mbedtls v3.x as crypto library */
+#define OPENTHREAD_CONFIG_CRYPTO_LIB_MBEDTLS_3 1
+/** Use ARM Platform Security Library as crypto library */
+#define OPENTHREAD_CONFIG_CRYPTO_LIB_PSA 2
+/** Use platform provided crypto library */
+#define OPENTHREAD_CONFIG_CRYPTO_LIB_PLATFORM 3
+
+#if OPENTHREAD_CONFIG_CRYPTO_LIB == OPENTHREAD_CONFIG_CRYPTO_LIB_PLATFORM
+
+/**
+ * @def OPENTHREAD_CONFIG_AES_CONTEXT_SIZE
+ *
+ * The size of the AES context byte array. Only applicable with OPENTHREAD_CONFIG_CRYPTO_LIB_PLATFORM.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_AES_CONTEXT_SIZE
+#define OPENTHREAD_CONFIG_AES_CONTEXT_SIZE (sizeof(void *))
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_HMAC_SHA256_CONTEXT_SIZE
+ *
+ * The size of the HMAC_SHA256 context byte array. Only applicable with OPENTHREAD_CONFIG_CRYPTO_LIB_PLATFORM.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_HMAC_SHA256_CONTEXT_SIZE
+#define OPENTHREAD_CONFIG_HMAC_SHA256_CONTEXT_SIZE (sizeof(void *))
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_HKDF_CONTEXT_SIZE
+ *
+ * The size of the HKDF context byte array. Only applicable with OPENTHREAD_CONFIG_CRYPTO_LIB_PLATFORM.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_HKDF_CONTEXT_SIZE
+#define OPENTHREAD_CONFIG_HKDF_CONTEXT_SIZE (sizeof(void *))
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_SHA256_CONTEXT_SIZE
+ *
+ * The size of the SHA256 context byte array. Only applicable with OPENTHREAD_CONFIG_CRYPTO_LIB_PLATFORM.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_SHA256_CONTEXT_SIZE
+#define OPENTHREAD_CONFIG_SHA256_CONTEXT_SIZE (sizeof(void *))
```

Comment:
How about adding `#error config so-and-so is missing` under `#indef` for these configs (instead of defining thme with  a default value of `sizeof(void*)`)?

I think if the `OPENTHREAD_CONFIG_CRYPTO_LIB == OPENTHREAD_CONFIG_CRYPTO_LIB_PLATFORM` then platform definitely needs to provides the context sizes and if they are not provided, it may be better to fail during build time and make it noticeable.

Reply:
I second @abtink's proposition.